Update beaver based generation of dmqnode docker image

daniel-zullo-sociomantic opened the pull request:
To better isolate and differentiate docker files and scripts.

matthias-wende-sociomantic commented:
All Travis instances fail, so maybe I was wrong and it is used to install libs.

leandro-lucarella-sociomantic commented:
It's the other way around. `docker/build` is used by beaver to build the project, see https://github.com/sociomantic-tsunami/beaver#beaver-install

leandro-lucarella-sociomantic commented:
The dhtnode is different because @mihails-strasuns is doing an extra isolation step to first build the node and then create a docker image with the binary IIRC. If there is not `docker/build` beaver will ignore it silently.

leandro-lucarella-sociomantic commented:
Wait a minute! I might have jumped into conclusions too early, the dmqnode is also doing the docker black magic, maybe I should check the exact error message first :D

daniel-zullo-sociomantic commented:
The `dhtnode` has a dedicated directory to build beaver. So the `build` script is needed but maybe we should do the same done for the `dhtnode` here (to distinguish the different builds)

leandro-lucarella-sociomantic commented:
There is a slight difference between both. The dhtnode is using a new beaver feature to isolate different dockerfiles better, the `docker/build` is actually there, but named `docker/builder/build` instead. See https://github.com/sociomantic-tsunami/dhtnode/blob/master/.travis.yml#L20. The dmq doesn't have that distinction, and thus use the default `docker/build` script to prepare the docker container to build the project.

daniel-zullo-sociomantic commented:
I pushes a fixup commit to do pretty much what I was done for the `dhtnode` it works fine. I'll amend the commits and issue name/description.

daniel-zullo-sociomantic commented:
Amended commit and issue title/description.

mihails-strasuns commented on docker/builder/beaver.Dockerfile:
```diff
@@ -0,0 +1 @@
+FROM sociomantictsunami/dlang:v4
```
Why not latest?

daniel-zullo-sociomantic replied:
will update

daniel-zullo-sociomantic replied:
Updated to `v6` as it is used in `docker/Dockerfile.dmqnode`

mihails-strasuns commented:
To clarify: reason why there were two different `build` scripts was that one was required by `beaver dlang install` command and used to run tests. Other was introduced when working on deployment image to be called from `Dockerfile.dmqnode`. Those are not interchangeable because latter has to do bunch of things beaver provides automatically for the former.

This duplication is annoying but I decided to ignore it considering the plan to move to gitlab because beaver will be gone there and build approach can be unified.

Switching to dhtnode style isolation is a good idea as it makes more obvious that these are independent systems.

daniel-zullo-sociomantic commented:
Updated based on a suggestion from Mihails.

<details>
<summary>Previous head was f6dcf69, new head at eb6a1fc. Diff between heads are:</summary>

```diff
diff --git a/beaver.Dockerfile b/beaver.Dockerfile
deleted file mode 100644
index 140287b..0000000
--- a/beaver.Dockerfile
+++ /dev/null
@@ -1 +0,0 @@
-FROM sociomantictsunami/dlang:v3
diff --git a/docker/builder/dockerignore b/docker/builder/.dockerignore
similarity index 100%
rename from docker/builder/dockerignore
rename to docker/builder/.dockerignore
diff --git a/docker/builder/beaver.Dockerfile b/docker/builder/beaver.Dockerfile
index 87a86da..f0b2b76 100644
--- a/docker/builder/beaver.Dockerfile
+++ b/docker/builder/beaver.Dockerfile
@@ -1 +1 @@
-FROM sociomantictsunami/dlang:v4
+FROM sociomantictsunami/dlang:latest
```
</details>

daniel-zullo-sociomantic commented:
`latest` didn't work 

<details>
<summary>Previous head was eb6a1fc, new head at 093743c. Diff between heads are:</summary>

```diff
diff --git a/.dockerignore b/.dockerignore
deleted file mode 100644
index 378eac2..0000000
--- a/.dockerignore
+++ /dev/null
@@ -1 +0,0 @@
-build
diff --git a/docker/builder/beaver.Dockerfile b/docker/builder/beaver.Dockerfile
index f0b2b76..98c0a6c 100644
--- a/docker/builder/beaver.Dockerfile
+++ b/docker/builder/beaver.Dockerfile
@@ -1 +1 @@
-FROM sociomantictsunami/dlang:latest
+FROM sociomantictsunami/dlang:v6
```
</details>

mihails-strasuns commented:
Yeah I did not mean it literally but latest current from https://hub.docker.com/r/sociomantictsunami/develdlang/tags - that would be `sociomantictsunami/develdlang:v7` right now

daniel-zullo-sociomantic commented:
> Yeah I did not mean it literally but latest current from https://hub.docker.com/r/sociomantictsunami/develdlang/tags - that would be sociomantictsunami/develdlang:v7 right now

Yes, I've just updated to that version

leandro-lucarella-sociomantic commented:
Travis doesn't approve, it's failing for dmd 2.078 with: [`./src/dmqnode/storage/engine/DiskOverflow.d(1535): Error: no property 'sort' for type 'uint[]'`](https://travis-ci.org/sociomantic-tsunami/dmqnode/jobs/457088985#L731).

matthias-wende-sociomantic commented:
I'll look into it, soon.

matthias-wende-sociomantic commented:
Cool, I did the same :), should have referenced it here. (https://github.com/sociomantic-tsunami/dmqnode/pull/77). I'll close my PR.

matthias-wende-sociomantic commented on src/dmqnode/storage/engine/DiskOverflow.d:
```diff
@@ -212,6 +212,7 @@ class DiskOverflow: DiskOverflowInfo
 
     import swarm.neo.protocol.socket.uio_const: iovec_const;
 
+    import ocean.core.Array: sort;
```
It may be worth mentioning this changelog https://dlang.org/changelog/2.075.0.html#removeArrayProps (pointed out by @Geod24 in https://github.com/sociomantic-tsunami/dmqnode/pull/77) in the commit.

daniel-zullo-sociomantic replied:
Right, I'll update the commit message.

daniel-zullo-sociomantic commented:
Amended the commit title and message and travis failed :s
